src/lib/bmp/bmp.c

```diff
@@ -0,0 +1,226 @@
+	read(fd,&(bmp_head->biSizeImage), 4);
+
+	read(fd,&(bmp_head->biXPelsPerMeter), 4);
+
+	read(fd,&(bmp_head->biYPelsPerMeter), 4);
+
+	read(fd,&(bmp_head->biClrUsed), 4);
+
+	read(fd,&(bmp_head->biClrImportant), 4);
+
+	return 0;
+}
+
+void print_bmp_head(bmp_header *bmp_head) {
+	printf("bfSize : 0x%x\nbfOffBits : 0x%x\nbiSize : 0x%x\n",\
+		bmp_head->bfSize, bmp_head->bfOffBits, bmp_head->biSize);
+	printf("biWidth : 0x%x\nbiHeight : 0x%x\nbiBitCount : 0x%x\n",\
+		bmp_head->biWidth, bmp_head->biHeight, bmp_head->biBitCount);
+	printf("biSizeImage : 0x%x\nbiXPelsPerMeter : 0x%x\nbiYPelsPerMeter : 0x%x\n",\
+		bmp_head->biSizeImage, bmp_head->biXPelsPerMeter, bmp_head->biYPelsPerMeter);
+	printf("biClrUsed : 0x%x\nbiClrImportant : 0x%x\n",\
+		bmp_head->biClrUsed, bmp_head->biClrImportant);
+}
+
+
+int  read_bmp(struct fb_var_screeninfo *vinfo, char *buf, char *file_name) {
+	int fd = 0 ;
+	int i, j, offset;
+	//unsigned int data = 0;
+	int addres = 0; 
+	unsigned char tmp[4] = {0};
+	int bmp_bytes_per_pixel, fb_bytes_per_pixel ;
+	bmp_header bmp_head;
+	color pallete[256];
+
+	fd = open((const char *)file_name, O_RDONLY);
+	if (fd < 0) {
+		perror("open");
+		exit(-1);
+	}
+
+	memset(&bmp_head, 0, sizeof(bmp_head));
+	read_bmp_head(fd, &bmp_head);
+	print_bmp_head(&bmp_head);
+	
+	if (bmp_head.biBitCount < 16) {
+		read_pallete(fd, pallete, (bmp_head.bfOffBits - 54) / 4);
+	}
+	lseek(fd, bmp_head.bfOffBits, 0); 
+
+	if ((bmp_head.biHeight > vinfo->yres) || (bmp_head.biWidth > vinfo->xres)) {
+		printf("Size of image bigger then resolution of fb\n");
+		return -1;
+	}
+
+	bmp_bytes_per_pixel = bmp_head.biBitCount / 8;
+	fb_bytes_per_pixel = vinfo->bits_per_pixel / 8;
+	/* TODO palette types */
+	/* TODO bad drawing some 24bit images */ 
+
+	for (i = bmp_head.biHeight-1; i>=0; i--) {
+			offset = 0;
+		for (j = bmp_head.biWidth * bmp_bytes_per_pixel - 1; j >= 0; j -= bmp_bytes_per_pixel){
+			addres = offset + i * vinfo->xres * fb_bytes_per_pixel; 
+			offset += fb_bytes_per_pixel;
+
+			switch(bmp_head.biBitCount) {
+				case 32:
+					memset(tmp, 0, bmp_bytes_per_pixel);
+					read(fd, tmp, bmp_bytes_per_pixel);
+
+					if (vinfo->bits_per_pixel == 32) {
+						* (buf+addres)   = tmp[0];
+						* (buf+addres+1) = tmp[1];
+						* (buf+addres+2) = tmp[2];
+						* (buf+addres+3) = 255;
+					}
+					break;
+				case 8:
+					memset(tmp, 0, bmp_bytes_per_pixel);
+					read(fd, tmp, bmp_bytes_per_pixel);
+
+					if (vinfo->bits_per_pixel == 32) {
+						* (buf+addres)   = pallete[tmp[0]].blue;
+						* (buf+addres+1) = pallete[tmp[0]].green;
+						* (buf+addres+2) = pallete[tmp[0]].red;
+						* (buf+addres+3) = 255;
+					}
+					break;
+			}
+		}
+	}
+	
+	return 0;
+}
+
+int draw_bmp(char *path) {
+	int fbfd = 0;
+	struct fb_var_screeninfo vinfo;
+	struct fb_fix_screeninfo finfo;
+	long int screensize = 0;
+	char *fbp = 0;
+
+	fbfd = open("/dev/fb0",O_RDWR);
+	if(fbfd == -1){
+		printf("Error : cannot open framebuffer device .\n");
+		exit(1);
```

Comment:
Why `exit()` instead of `return`? This will kill the process which calls this function. Probably that's not what user expects.

Reply:
done